File: src/HomeOS.API/Modules/Shopping/Application/ShoppingListService.cs
using HomeOS.API.Modules.Products.Infrastructure;
using HomeOS.API.Modules.Shopping.Application.DTOs;
using HomeOS.API.Modules.Shopping.Domain;
using HomeOS.API.Modules.Shopping.Infrastructure;

namespace HomeOS.API.Modules.Shopping.Application;

public class ShoppingListService
{
    private readonly IShoppingListRepository _shoppingListRepository;
    private readonly IProductRepository _productRepository;

    public ShoppingListService(
        IShoppingListRepository shoppingListRepository,
        IProductRepository productRepository)
    {
        _shoppingListRepository = shoppingListRepository;
        _productRepository = productRepository;
    }

    public async Task<ShoppingListResponse> CreateAsync(CreateShoppingListRequest request, Guid householdId)
    {
        var list = ShoppingList.Create(householdId, request.Name);
        await _shoppingListRepository.CreateAsync(list);
        return await MapToResponseAsync(list);
    }

    public async Task<ShoppingListResponse?> GetByIdAsync(Guid id, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        return list is null ? null : await MapToResponseAsync(list);
    }

    public async Task<IEnumerable<ShoppingListSummaryResponse>> GetAllAsync(Guid householdId)
    {
        var lists = await _shoppingListRepository.GetAllAsync(householdId);
        var summaries = new List<ShoppingListSummaryResponse>();
        foreach (var list in lists)
        {
            // Reload list with items to calculate summary
            var fullList = await _shoppingListRepository.FindByIdAsync(list.Id, householdId);
            if (fullList != null)
            {
                summaries.Add(new ShoppingListSummaryResponse(
                    fullList.Id,
                    fullList.HouseholdId,
                    fullList.Name,
                    fullList.Status.ToString(),
                    fullList.CreatedAt,
                    fullList.UpdatedAt,
                    fullList.Items.Count,
                    fullList.Items.Count(i => i.Checked)
                ));
            }
        }
        return summaries;
    }

    public async Task<ShoppingListResponse> UpdateAsync(Guid id, UpdateShoppingListRequest request, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        list.UpdateName(request.Name);
        await _shoppingListRepository.UpdateAsync(list);
        return await MapToResponseAsync(list);
    }

    public async Task DeleteAsync(Guid id, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        await _shoppingListRepository.DeleteAsync(id, householdId);
    }

    public async Task<ShoppingListResponse> StartShoppingAsync(Guid id, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        list.StartShopping();
        await _shoppingListRepository.UpdateAsync(list);
        return await MapToResponseAsync(list);
    }

    public async Task<ShoppingListResponse> CompleteShoppingAsync(Guid id, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        list.CompleteShopping();
        await _shoppingListRepository.UpdateAsync(list);
        return await MapToResponseAsync(list);
    }

    public async Task<ShoppingListResponse> CancelShoppingAsync(Guid id, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(id, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        list.CancelShopping();
        await _shoppingListRepository.UpdateAsync(list);
        return await MapToResponseAsync(list);
    }

    public async Task<ShoppingListItemResponse> AddItemAsync(Guid listId, AddItemRequest request, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(listId, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        var product = await _productRepository.FindByIdAsync(request.ProductId, householdId);
        if (product is null)
            throw new KeyNotFoundException("Produto não encontrado.");

        var item = list.AddItem(request.ProductId, request.Quantity, request.Unit);
        
        await _shoppingListRepository.UpdateAsync(list); // Updates list timestamp
        await _shoppingListRepository.CreateItemAsync(item);
        
        return new ShoppingListItemResponse(
            item.Id,
            item.ProductId,
            product.Name,
            product.Brand,
            product.Barcode,
            item.Quantity,
            item.Unit,
            item.Checked,
            item.CreatedAt,
            item.UpdatedAt);
    }

    public async Task RemoveItemAsync(Guid listId, Guid itemId, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(listId, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        list.RemoveItem(itemId); // Domain logic validation
        
        await _shoppingListRepository.UpdateAsync(list);
        await _shoppingListRepository.DeleteItemAsync(itemId, listId);
    }

    public async Task<ShoppingListItemResponse> CheckItemAsync(Guid listId, Guid itemId, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(listId, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        var item = await _shoppingListRepository.FindItemByIdAsync(itemId, listId);
        if (item is null)
            throw new KeyNotFoundException("Item não encontrado.");

        item.Check();
        await _shoppingListRepository.UpdateItemAsync(item);

        return await MapItemToResponseAsync(item, householdId);
    }

    public async Task<ShoppingListItemResponse> UncheckItemAsync(Guid listId, Guid itemId, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(listId, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        var item = await _shoppingListRepository.FindItemByIdAsync(itemId, listId);
        if (item is null)
            throw new KeyNotFoundException("Item não encontrado.");

        item.Uncheck();
        await _shoppingListRepository.UpdateItemAsync(item);

        return await MapItemToResponseAsync(item, householdId);
    }
    
    public async Task<ShoppingListItemResponse> UpdateItemQuantityAsync(Guid listId, Guid itemId, UpdateItemQuantityRequest request, Guid householdId)
    {
        var list = await _shoppingListRepository.FindByIdAsync(listId, householdId);
        if (list is null)
            throw new KeyNotFoundException("Lista não encontrada.");

        var item = await _shoppingListRepository.FindItemByIdAsync(itemId, listId);
        if (item is null)
            throw new KeyNotFoundException("Item não encontrado.");

        item.UpdateQuantity(request.Quantity);
        await _shoppingListRepository.UpdateItemAsync(item);

        return await MapItemToResponseAsync(item, householdId);
    }

    private async Task<ShoppingListResponse> MapToResponseAsync(ShoppingList list)
    {
        var itemResponses = new List<ShoppingListItemResponse>();
        foreach (var item in list.Items)
        {
            var product = await _productRepository.FindByIdAsync(item.ProductId, list.HouseholdId);
            if (product != null)
            {
                itemResponses.Add(new ShoppingListItemResponse(
                    item.Id,
                    item.ProductId,
                    product.Name,
                    product.Brand,
                    product.Barcode,
                    item.Quantity,
                    item.Unit,
                    item.Checked,
                    item.CreatedAt,
                    item.UpdatedAt));
            }
        }

        return new ShoppingListResponse(
            list.Id,
            list.HouseholdId,
            list.Name,
            list.Status.ToString(),
            list.CreatedAt,
            list.UpdatedAt,
            itemResponses);
    }
    
    private async Task<ShoppingListItemResponse> MapItemToResponseAsync(ShoppingListItem item, Guid householdId)
    {
        var product = await _productRepository.FindByIdAsync(item.ProductId, householdId);
        if (product == null) throw new InvalidOperationException("Product not found");
        
        return new ShoppingListItemResponse(
            item.Id,
            item.ProductId,
            product.Name,
            product.Brand,
            product.Barcode,
            item.Quantity,
            item.Unit,
            item.Checked,
            item.CreatedAt,
            item.UpdatedAt);
    }
}

File: src/HomeOS.Web/src/app/core/shopping-list.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export type ShoppingListStatus = 'Pending' | 'InProgress' | 'Completed' | 'Cancelled';

export interface ShoppingListItem {
  id: string;
  productId: string;
  productName: string;
  productBrand?: string;
  productBarcode?: string;
  quantity: number;
  unit?: string;
  price?: number;
  checked: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ShoppingList {
  id: string;
  householdId: string;
  name: string;
  status: ShoppingListStatus;
  createdAt: string;
  updatedAt: string;
  items: ShoppingListItem[];
}

export interface ShoppingListSummary {
  id: string;
  householdId: string;
  name: string;
  status: ShoppingListStatus;
  createdAt: string;
  updatedAt: string;
  totalItems: number;
  checkedItems: number;
}

@Injectable({
  providedIn: 'root'
})
export class ShoppingListService {
  private http = inject(HttpClient);
  private baseUrl = '/api/shopping-lists';

  getLists(householdId: string): Observable<ShoppingListSummary[]> {
    return this.http.get<ShoppingListSummary[]>(`${this.baseUrl}?householdId=${householdId}`);
  }

  getListById(id: string, householdId: string): Observable<ShoppingList> {
    return this.http.get<ShoppingList>(`${this.baseUrl}/${id}?householdId=${householdId}`);
  }

  createList(name: string, householdId: string): Observable<ShoppingList> {
    return this.http.post<ShoppingList>(`${this.baseUrl}?householdId=${householdId}`, { name });
  }

  deleteList(id: string, householdId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}?householdId=${householdId}`);
  }

  startShopping(id: string, householdId: string): Observable<ShoppingList> {
    return this.http.post<ShoppingList>(`${this.baseUrl}/${id}/start?householdId=${householdId}`, {});
  }

  completeShopping(id: string, householdId: string): Observable<ShoppingList> {
    return this.http.post<ShoppingList>(`${this.baseUrl}/${id}/complete?householdId=${householdId}`, {});
  }

  addItem(listId: string, productId: string, quantity: number, householdId: string, unit?: string, price?: number): Observable<ShoppingListItem> {
    return this.http.post<ShoppingListItem>(`${this.baseUrl}/${listId}/items?householdId=${householdId}`, { productId, quantity, unit, price });
  }

  removeItem(listId: string, itemId: string, householdId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${listId}/items/${itemId}?householdId=${householdId}`);
  }

  checkItem(listId: string, itemId: string, householdId: string): Observable<ShoppingListItem> {
    return this.http.put<ShoppingListItem>(`${this.baseUrl}/${listId}/items/${itemId}/check?householdId=${householdId}`, {});
  }

  uncheckItem(listId: string, itemId: string, householdId: string): Observable<ShoppingListItem> {
    return this.http.put<ShoppingListItem>(`${this.baseUrl}/${listId}/items/${itemId}/uncheck?householdId=${householdId}`, {});
  }

  updateItemDetails(listId: string, itemId: string, quantity: number, householdId: string, price?: number): Observable<ShoppingListItem> {
    return this.http.put<ShoppingListItem>(`${this.baseUrl}/${listId}/items/${itemId}/details?householdId=${householdId}`, { quantity, price });
  }
}
